xrdp/xrdp_mm.c

```diff
@@ -1328,33 +1309,54 @@ dynamic_monitor_process_queue(struct xrdp_mm *self)
         return 0;
     }
 
-    while (self->resize_queue->count > 0)
+    description = &self->wm->mm->resize_data;
+
+    if (*description == NULL)
     {
-        description = (struct dynamic_monitor_description *)
-                      list_get_item(self->resize_queue, 0);
-        if (description->state == WMRZ_COMPLETE)
-        {
-            LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Clearing"
-                " completed resize (w: %d x h: %d). It took %d milliseconds.",
-                description->description.session_width,
-                description->description.session_height,
-                g_time3() - description->start_time);
-        }
-        else if (description->state == WMRZ_ERROR)
+        if  (self->resize_queue->count > 0)
         {
-            LOG(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Clearing"
-                " failed request to resize to: (w: %d x h: %d)",
-                description->description.session_width,
-                description->description.session_height);
-        }
-        else
-        {
-            // No errors, process it!
-            return process_dynamic_monitor_description(self->wm, description);
+            LOG_DEVEL(LOG_LEVEL_INFO, "dynamic_monitor_process_queue: Queue is"
+                      "not empty. Filling out description.");
+            const tintptr queue_head = list_get_item(self->resize_queue, 0);
+            (*description) = (struct display_control_monitor_layout_data *)
+                             g_malloc(sizeof(struct display_control_monitor_layout_data), 1);
+
+            g_memcpy(&((*description)->description),
+                     (struct display_size_description *)queue_head,
+                     sizeof(struct display_size_description));
+            (*description)->start_time = g_time3();
+            advance_resize_state_machine(self->wm->mm,
+                                         *description,
+                                         WMRZ_ENCODER_DELETE);
+            list_remove_item(self->resize_queue, 0);
+            g_free((struct display_size_description *)queue_head);
```

Comment:
If you set `auto_free` for the resize_queue, you can remove this `g_free()` call as `list_remove_item()` will do it for you. That will also fix the possible memory leak in `xrdp_mm_delete()` where you are calling `list_delete()` on the `resize_queue` without checking it's empty first.

Reply:
Updated.